spotbugs/src/main/java/edu/umd/cs/findbugs/detect/MultipleInstantiationsOfSingletons.java

```diff
@@ -0,0 +1,225 @@
+                    cloneOnlyThrowsCloneNotSupportedException = true;
+                }
+            }
+        } else if (seen == Const.ARETURN) {
+            OpcodeStack.Item item = stack.getStackItem(0);
+            XField field = item.getXField();
+            if (field != null) {
+                String className = "L" + getClassName() + ";";
+                if (field.isPrivate() && field.isStatic() && field.getSignature().equals(className)) {
+                    isSingleton = true;
+                    isGetterMethodSynchronized = getMethod().isSynchronized();
+                    methods.put(Methods.INSTANCE_GETTER, getXMethod());
+                }
+            }
+        } else if (seen == Const.MONITORENTER) {
+            methodsUsingMonitor.add(getXMethod());
+        }
+    }
+
+    @Override
+    public void visitAfter(JavaClass obj) {
+        if (!isSingleton) {
+            return;
+        }
+
+        boolean hasNonPrivateConstructor = false;
+        for (XMethod constructor : constructors) {
+            if (!constructor.isPrivate()) {
+                hasNonPrivateConstructor = true;
+                methods.put(Methods.CONSTRUCTOR, constructor);
+                break;
+            }
+        }
+
+        boolean isGetterMethodUsingMonitor = methodsUsingMonitor.contains(methods.get(Methods.INSTANCE_GETTER));
+
+        if (hasNonPrivateConstructor) {
+            bugReporter.reportBug(new BugInstance(this, "SING_SINGLETON_HAS_NONPRIVATE_CONSTRUCTOR", NORMAL_PRIORITY).addClass(this).addMethod(
+                    methods.get(Methods.CONSTRUCTOR)));
+        }
+
+        if (!isGetterMethodSynchronized && !isGetterMethodUsingMonitor) {
+            bugReporter.reportBug(new BugInstance(this, "SING_SINGLETON_GETTER_NOT_SYNCHRONIZED", NORMAL_PRIORITY).addClass(this).addMethod(
+                    methods.get(Methods.INSTANCE_GETTER)));
+        }
+
+        if (isCloneable) {
+            if (implementsCloneableDirectly) { // directly
+                bugReporter.reportBug(new BugInstance(this, "SING_SINGLETON_IMPLEMENTS_CLONEABLE", NORMAL_PRIORITY).addClass(this).addMethod(methods
+                        .get(Methods.CLONE)));
+            } else { // indirectly
+                if (!cloneOnlyThrowsCloneNotSupportedException) { // no or not only CloneNotSupportedException
+                    bugReporter.reportBug(new BugInstance(this, "SING_SINGLETON_INDIRECTLY_IMPLEMENTS_CLONEABLE", NORMAL_PRIORITY).addClass(this)
+                            .addMethod(methods.get(Methods.CLONE)));
+                }
+            }
+        } else if (hasCloneMethod) {
+            bugReporter.reportBug(new BugInstance(this, "SING_SINGLETON_IMPLEMENTS_CLONE_METHOD", NORMAL_PRIORITY).addClass(this).addMethod(methods
+                    .get(Methods.CLONE)));
+        }
+
+        if (isSerializable) {
+            bugReporter.reportBug(new BugInstance(this, "SING_SINGLETON_IMPLEMENTS_SERIALIZABLE", NORMAL_PRIORITY).addClass(this));
```

Comment:
[This is accepted in this case](https://wiki.sei.cmu.edu/confluence/display/java/MSC07-J.+Prevent+multiple+instantiations+of+singleton+objects#MSC07J.Preventmultipleinstantiationsofsingletonobjects-CompliantSolution(Overrideclone()Method)#MSC07J.Preventmultipleinstantiationsofsingletonobjects-CompliantSolution(EnumerationTypes)).
Please add a test case which uses the example class in that doc:

```
public enum MySingleton {
  ; // Empty list of enum values
 
  private static MySingleton instance;
 
  // Nontransient field
  private String[] str = {"one", "two", "three"};
 
  public void displayStr() {
    System.out.println(Arrays.toString(str));
  } 
}
```

Reply:
There was a testcase like this added, so I'm resolving this conversation.